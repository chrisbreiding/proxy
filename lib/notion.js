const rp = require('request-promise')

const getData = ({ notionPageId, notionToken }) => {
  return rp({
    uri: `https://api.notion.com/v1/blocks/${notionPageId}/children`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

module.exports = {
  getData,
}
