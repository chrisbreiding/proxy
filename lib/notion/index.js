const { addUpcomingWeek } = require('./upcoming-week')
const { getData } = require('./data')
const { onSocket } = require('./shopping')

module.exports = {
  addUpcomingWeek,
  getData,
  onSocket,
}
