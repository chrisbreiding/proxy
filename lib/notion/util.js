const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
