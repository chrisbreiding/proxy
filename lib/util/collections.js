const compact = (array) => {
  return array.reduce((memo, item) => {
    return item ? memo.concat(item) : memo
  }, [])
}

module.exports = {
  compact,
}
