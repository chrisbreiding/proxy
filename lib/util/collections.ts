export function clone<T> (obj: T): T {
  return JSON.parse(JSON.stringify(obj))
}

export function compact<T> (array: (T | undefined)[]): T[] {
  return array.reduce((memo: T[], item: T | undefined) => {
    return item ? memo.concat(item) : memo
  }, [] as T[])
}

export async function mapPromisesSerially<T, U> (array: U[], callback: (arg: U) => Promise<T>) {
  const result: T[] = []

  for (const item of array) {
    result.push(await callback(item))
  }

  return result
}

export function times (num: number, filler?: any) {
  return (new Array(num)).fill(filler)
}
