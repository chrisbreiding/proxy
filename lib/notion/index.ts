import { addUpcomingWeek, upcomingWeekView } from './upcoming-week'
import { getBlockChildren } from './util'
import { onSocket } from './shopping'

interface GetDataOptions {
  notionToken: string
  notionPageId: string
}

export function getNotionData ({ notionToken, notionPageId }: GetDataOptions) {
  return getBlockChildren({ notionToken, pageId: notionPageId })
}

export {
  addUpcomingWeek,
  upcomingWeekView,
  onSocket,
}
