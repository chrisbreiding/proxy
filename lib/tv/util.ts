import type { Method } from 'axios'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { request } from '../util/network'

const apikey = getEnv('THETVDB_API_KEY')
const pin = getEnv('THETVDB_PIN')
export const baseUrl = 'https://api4.thetvdb.com'

interface LoginResult {
  status: string
  data: {
    token: string
  }
}

async function authenticate () {
  debugVerbose('authenticate')

  try {
    const { data, status } = await request({
      method: 'post',
      url: `${baseUrl}/v4/login`,
      body: { apikey, pin },
    }) as LoginResult

    if (!data?.token) {
      throw new Error(`Could not authenticate with TheTVDB, status: ${status}`)
    }

    return data.token as string
  } catch (error: any) {
    debug('Authentication failed:', error?.stack || error)

    throw error
  }
}

interface MakeRequestOptions {
  method?: Method
  params?: { [key: string]: string }
  path: string
  token?: string
}

export async function makeRequest (options: MakeRequestOptions) {
  const { method, params, path } = options
  const token = options.token || await authenticate()
  const url = `${baseUrl}/v4/${path}`

  debugVerbose('request: %o', { method, url, params, token })

  // TODO: handle error
  const result = await request({
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    method,
    params,
    url,
  })

  result.token = token

  return result
}
