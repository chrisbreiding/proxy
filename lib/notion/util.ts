import type { Method } from 'axios'
import dayjs from 'dayjs'
import type {
  ApiColor,
  BlockObjectResponse,
  BookmarkBlockObjectResponse,
  BulletedListItemBlockObjectResponse,
  CalloutBlockObjectResponse,
  ChildDatabaseBlockObjectResponse,
  ChildPageBlockObjectResponse,
  ColumnBlockObjectResponse,
  ColumnListBlockObjectResponse,
  DividerBlockObjectResponse,
  EmbedBlockObjectResponse,
  EquationBlockObjectResponse,
  FileBlockObjectResponse,
  Heading1BlockObjectResponse,
  Heading2BlockObjectResponse,
  Heading3BlockObjectResponse,
  ImageBlockObjectResponse,
  LinkPreviewBlockObjectResponse,
  LinkToPageBlockObjectResponse,
  ListBlockChildrenResponse,
  ListDatabasesResponse,
  NumberedListItemBlockObjectResponse,
  ParagraphBlockObjectResponse,
  PdfBlockObjectResponse,
  QuoteBlockObjectResponse,
  RichTextItemResponse,
  SyncedBlockBlockObjectResponse,
  TableBlockObjectResponse,
  TableOfContentsBlockObjectResponse,
  TableRowBlockObjectResponse,
  TemplateBlockObjectResponse,
  ToDoBlockObjectResponse,
  ToggleBlockObjectResponse,
  UnsupportedBlockObjectResponse,
  UpdateBlockBodyParameters,
  VideoBlockObjectResponse,
} from '@notionhq/client/build/src/api-endpoints'

import { clone, compact } from '../util/collections'
import { debug } from '../util/debug'
import { request } from '../util/network'

export const dateRegex = /(?:Sun|Mon|Tue|Wed|Thu|Fri|Sat), (\d{1,2}\/\d{1,2})/

export function getDateFromText (dateText: string) {
  const currentDate = dayjs()
  // originally assume the date's year matches the current year
  const date = dayjs(`${dateText}/${currentDate.year()}`, 'M/D/YYYY')
  // if the current month is after the date's month, we've crossed over years
  // and the date's year should be the next year. usually, it's because it's
  // currently December, but the date's month is January. this can be
  // generically applied, so it will work if it's currently June, but we're
  // trying to get the weather for April. we don't want past weather, only
  // future weather, so assume it's for the following year.
  if (currentDate.month() > date.month()) {
    return date.add(1, 'year')
  }

  return date
}

interface MakeRequestOptions {
  body?: object
  method?: Method
  notionToken: string
  path: string
}

function makeRequest<T> (options: MakeRequestOptions): Promise<T> {
  const { notionToken, body, path, method = 'get' } = options

  return request({
    method,
    body,
    url: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2022-06-28',
    },
  })
}

interface GetBlockChildrenOptions {
  notionToken: string
  pageId: string
}

export async function getBlockChildren ({ notionToken, pageId }: GetBlockChildrenOptions) {
  const { results } = await makeRequest<ListBlockChildrenResponse>({ notionToken, path: `blocks/${pageId}/children` }) as { results: BlockObjectResponse[] }

  return results.map(convertNotionBlockToOwnBlock)
}

interface GetBlockChildrenDeepOptions {
  filter?: (block: NotionBlock) => boolean
  includeId?: boolean
  notionToken: string
  pageId: string
}

export async function getBlockChildrenDeep ({ notionToken, pageId, filter, includeId }: GetBlockChildrenDeepOptions): Promise<NotionBlock[]> {
  const blocks = await getBlockChildren({ notionToken, pageId })
  const filteredBlocks = filter ? blocks.filter(filter) : blocks
  const blocksWithChildren = await Promise.all(filteredBlocks.map(async (block) => {
    if (block.has_children && block.type !== 'child_page') {
      block.children = await getBlockChildrenDeep({ notionToken, pageId: block.id, includeId })
    }

    return block
  }))

  return compact(blocksWithChildren)
}

type Content = BookmarkBlockObjectResponse['bookmark']
| BulletedListItemBlockObjectResponse['bulleted_list_item']
| CalloutBlockObjectResponse['callout']
| ChildDatabaseBlockObjectResponse['child_database']
| ChildPageBlockObjectResponse['child_page']
| ColumnBlockObjectResponse['column']
| ColumnListBlockObjectResponse['column_list']
| DividerBlockObjectResponse['divider']
| EmbedBlockObjectResponse['embed']
| EquationBlockObjectResponse['equation']
| FileBlockObjectResponse['file']
| Heading1BlockObjectResponse['heading_1']
| Heading2BlockObjectResponse['heading_2']
| Heading3BlockObjectResponse['heading_3']
| ImageBlockObjectResponse['image']
| LinkPreviewBlockObjectResponse['link_preview']
| LinkToPageBlockObjectResponse['link_to_page']
| NumberedListItemBlockObjectResponse['numbered_list_item']
| ParagraphBlockObjectResponse['paragraph']
| PdfBlockObjectResponse['pdf']
| QuoteBlockObjectResponse['quote']
| SyncedBlockBlockObjectResponse['synced_block']
| TableBlockObjectResponse['table']
| TableOfContentsBlockObjectResponse['table_of_contents']
| TableRowBlockObjectResponse['table_row']
| TemplateBlockObjectResponse['template']
| ToDoBlockObjectResponse['to_do']
| ToggleBlockObjectResponse['toggle']
| UnsupportedBlockObjectResponse['unsupported']
| VideoBlockObjectResponse['video']

export interface OwnBlock {
  children?: OwnBlock[] | NotionBlock[]
  type: BlockObjectResponse['type']
  content: Content
}

export interface NotionBlock extends OwnBlock {
  children?: NotionBlock[]
  has_children: boolean
  id: string
}

export type Block = OwnBlock | NotionBlock

function getRichText (block: OwnBlock): RichTextItemResponse[] | undefined {
  return 'rich_text' in block.content ? block.content.rich_text : undefined
}

interface MakeAppendRequestOptions {
  blocks: OutgoingBlock[]
  notionToken: string
  pageId: string
}

export function makeAppendRequest ({ notionToken, pageId, blocks }: MakeAppendRequestOptions) {
  return makeRequest<ListBlockChildrenResponse>({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface AppendBlockChildrenOptions {
  blocks: OwnBlock[]
  notionToken: string
  pageId: string
}

export async function appendBlockChildrenDeep ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  function moveChildren (blocks: NotionBlock[] | OwnBlock[]) {
    return blocks.map((block) => {
      const convertedBlock = convertBlockToOutgoingBlock(block)

      if (!block.children) return convertedBlock

      // @ts-ignore
      convertedBlock[block.type].children = moveChildren(block.children)

      return convertedBlock
    })
  }

  return makeAppendRequest({ notionToken, pageId, blocks: moveChildren(blocks) })
}

export async function appendBlockChildren ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  let toAppend: OutgoingBlock[] = []

  for (const block of blocks) {
    const convertedBlock = convertBlockToOutgoingBlock(block)

    if (block.children) {
      const children = block.children

      toAppend.push(convertedBlock)

      const { results } = await makeAppendRequest({ notionToken, pageId, blocks: toAppend })
      toAppend = []

      await appendBlockChildren({
        notionToken,
        pageId: results[results.length - 1].id,
        blocks: children,
      })
    } else {
      toAppend.push(convertedBlock)
    }
  }

  if (toAppend.length) {
    return makeAppendRequest({
      notionToken,
      pageId,
      blocks: toAppend,
    })
  }
}

interface UpdateBlockOptions {
  notionToken: string
  blockId: string
  block: Block
}

export function updateBlock ({ notionToken, blockId, block }: UpdateBlockOptions) {
  const convertedBlock = convertBlockToOutgoingBlock(block)

  return makeRequest<void>({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: convertedBlock,
  })
}

interface UpdatePageOptions {
  notionToken: string
  pageId: string
  properties: {
    [key: string]: {
      date: {
        start: string
      }
    }
  }
}

export function updatePage ({ notionToken, pageId, properties }: UpdatePageOptions) {
  return makeRequest<void>({
    notionToken,
    body: { properties },
    method: 'patch',
    path: `pages/${pageId}`,
  })
}

interface QueryDatabasesOptions {
  notionToken: string
  databaseId: string
}

export function queryDatabases ({ notionToken, databaseId }: QueryDatabasesOptions) {
  return makeRequest<ListDatabasesResponse>({
    notionToken,
    method: 'post',
    path: `databases/${databaseId}/query`,
  })
}

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type

  return {
    // @ts-ignore
    content: clone(block[type] as Content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

type OutgoingBlock = UpdateBlockBodyParameters & {
  object: string
  type: BlockObjectResponse['type']
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface MakeBlockOptions {
  annotations?: Partial<RichTextItemResponse['annotations']>
  children?: OwnBlock[] | NotionBlock[]
  text: string
  type: BlockObjectResponse['type']
}

export function makeBlock ({ text, type = 'paragraph', annotations, children }: MakeBlockOptions) {
  return {
    type,
    children,
    content: {
      rich_text: [
        {
          type: 'text',
          text: {
            content: text,
          },
          annotations,
          plain_text: text,
        },
      ],
    },
  } as OwnBlock
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

export function getBlockPlainText (block: Block) {
  const richText = getRichText(block)

  if (!richText) return

  return richTextToPlainText(richText)
}

interface RichTextItem {
  text: {
    content: string
    link?: { url: string } | null
  }
  type?: 'text'
  annotations?: {
    bold?: boolean
    italic?: boolean
    strikethrough?: boolean
    underline?: boolean
    code?: boolean
    color?: ApiColor
  }
}

export function makeTextPart (content: string, color?: ApiColor) {
  const textPart = { text: { content } } as RichTextItem

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

export function getMonthNameFromIndex (monthIndex: number, short = false) {
  return dayjs().month(monthIndex).format(short ? 'MMM' : 'MMMM')
}

export function getMonths ({ short }: { short?: boolean } = {}) {
  return Array.from(new Array(12)).map((_, i) => {
    return getMonthNameFromIndex(i, short)
  })
}

/* c8 ignore start */
const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
/* c8 ignore stop */
