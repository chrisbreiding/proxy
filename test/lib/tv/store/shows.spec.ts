import nock from 'nock'
import { beforeEach, describe, expect, it, vi } from 'vitest'

const apikey = process.env.THETVDB_API_KEY = 'api-key'
const pin = process.env.THETVDB_PIN = 'pin'

import { baseUrl } from '../../../../lib/tv/source/util'
import { fixtureContents } from '../../notion/util'
import { nockLogin } from '../util'
import { startServer } from '../../../../index'
import { handleServer } from '../../../support/util'
import {
  addCollectionToDoc,
  addDoc,
  deleteDoc,
  getDoc,
  getDocWhere,
  updateDoc,
} from '../../../../lib/tv/store/firebase'
import { clone } from '../../../../lib/util/collections'

function makeSearchResultShow (num: number) {
  return {
    description: `Show ${num} description`,
    firstAired: `Show ${num} firstAired`,
    id: `show-${num}-id`,
    name: `Show ${num} name`,
    network: `Show ${num} network`,
    poster: `Show ${num} poster`,
    status: 'Ended' as const,
  }
}

function makeEpisode (num: number) {
  return {
    title: `Episode ${num}`,
  }
}

interface ShowUsers {
  [key: string]: {
    displayName: string
    fileName: string
    searchName: string
  }
}

function makeShow (num: number, episodes: any[], users: number[]) {
  return {
    episodes,
    id: `show-${num}-id`,
    name: `Show ${num} name`,
    users: users.reduce((memo, userNum) => {
      return {
        ...memo,
        [`user-${userNum}`]: {
          displayName: `user-${userNum} display name`,
          fileName: `user-${userNum} file name`,
          searchName: `user-${userNum} search name`,
        },
      }
    }, {} as ShowUsers),
  }
}

vi.mock('../../../../lib/tv/store/firebase', () => {
  return {
    addDoc: vi.fn(),
    addCollectionToDoc: vi.fn(),
    getCollection: vi.fn().mockResolvedValue([
      makeShow(1, [
        makeEpisode(1),
        makeEpisode(2),
      ], [1]),
      makeShow(2, [
        makeEpisode(3),
        makeEpisode(4),
      ], [1, 2]),
      makeShow(3, [
        makeEpisode(5),
        makeEpisode(6),
      ], [2]),
    ]),
    getDoc: vi.fn(),
    getDocWhere: vi.fn(),
    deleteDoc: vi.fn(),
    updateDoc: vi.fn(),
  }
})

describe('lib/tv/store/shows', () => {
  handleServer(startServer)

  beforeEach(() => {
    vi.clearAllMocks()
  })

  describe('user validation', () => {
    it('sends 401 if no api key', async (ctx) => {
      // @ts-ignore
      getDocWhere.mockResolvedValue(undefined)

      const res = await ctx.request.get('/tv/shows')

      expect(res.status).to.equal(401)
      expect(res.body).to.deep.equal({ error: 'Must specify `api-key` header' })
    })

    it('sends 401 if api key does not match any users', async (ctx) => {
      const res = await ctx.request.get('/tv/shows').set('api-key', 'no-match')

      expect(res.status).to.equal(401)
      expect(res.body).to.deep.equal({ error: 'Could not find user with api key: no-match' })
    })
  })

  describe('GET /tv/shows', () => {
    it('returns shows from store for user', async (ctx) => {
      // @ts-ignore
      getDocWhere.mockResolvedValue({ id: 'user-1' })

      const res = await ctx.request.get('/tv/shows').set('api-key', 'user-1-api-key')

      expect(res.status).to.equal(200)
      expect(res.body).toMatchSnapshot()
    })
  })

  describe('POST /tv/shows', () => {
    it('gets the show data, saves it, and adds the show to the user if it does not already exist', async (ctx) => {
      nockLogin(apikey, pin)

      nock(baseUrl)
      .get('/v4/series/show-2-id/episodes/default?page=0')
      .reply(200, fixtureContents('tv/episodes'))

      // @ts-ignore
      getDocWhere.mockResolvedValue({ id: 'user-1' })
      // @ts-ignore
      getDoc.mockResolvedValue(undefined)

      const searchShow = makeSearchResultShow(2)
      const res = await ctx.request.post('/tv/shows')
      .set('api-key', 'user-1-api-key')
      .send({ show: searchShow })

      expect(addDoc).toBeCalledWith(`shows/${searchShow.id}`, {
        id: searchShow.id,
        name: searchShow.name,
        poster: searchShow.poster,
        status: searchShow.status,
        users: {
          'user-1': {
            displayName: searchShow.name,
            fileName: searchShow.name,
            searchName: searchShow.name,
          },
        },
      })
      expect(addCollectionToDoc).toBeCalledWith(`shows/${searchShow.id}/episodes`, expect.arrayContaining([{
        airdate: '2009-02-17T05:00:00.000Z',
        episodeNumber: 1,
        season: 0,
        sourceId: 3859781,
        title: 'Good Cop Bad Cop',
      }]))
      expect(res.status).to.equal(200)
      expect(res.body).toMatchSnapshot()
    })

    it('adds the show to the user if it already exists', async (ctx) => {
      const show = makeShow(3, [makeEpisode(1)], [2])

      // @ts-ignore
      getDocWhere.mockResolvedValue({ id: 'user-1' })
      // @ts-ignore
      getDoc.mockResolvedValue(show)

      const searchShow = makeSearchResultShow(3)
      const res = await ctx.request.post('/tv/shows')
      .set('api-key', 'user-1-api-key')
      .send({ show: searchShow })

      const users = clone(show.users)

      users['user-1'] = {
        displayName: searchShow.name,
        fileName: searchShow.name,
        searchName: searchShow.name,
      }

      expect(updateDoc).toBeCalledWith(`shows/${searchShow.id}`, { users })
      expect(res.status).to.equal(200)
      expect(res.body).toMatchSnapshot()
    })

    it('send 204 status if show already exist and belongs to user', async (ctx) => {
      const show = makeShow(1, [makeEpisode(1)], [1])

      // @ts-ignore
      getDocWhere.mockResolvedValue({ id: 'user-1' })
      // @ts-ignore
      getDoc.mockResolvedValue(show)

      const searchShow = makeSearchResultShow(3)
      const res = await ctx.request.post('/tv/shows')
      .set('api-key', 'user-1-api-key')
      .send({ show: searchShow })

      expect(res.status).to.equal(204)
    })
  })

  describe('PUT /tv/shows/:id', () => {
    it('updates the show and returns it', async (ctx) => {
      const show = makeShow(1, [makeEpisode(1)], [1])

      // @ts-ignore
      getDocWhere.mockResolvedValue({ id: 'user-1' })
      // @ts-ignore
      getDoc.mockResolvedValue(show)

      const showUpdate = {
        displayName: 'new display name',
        fileName: 'new file name',
        searchName: 'new search name',
      }

      const res = await ctx.request.put(`/tv/shows/${show.id}`)
      .set('api-key', 'user-1-api-key')
      .send({ show: showUpdate })

      expect(updateDoc).toBeCalledWith(`shows/${show.id}`, {
        id: 'show-1-id',
        name: 'Show 1 name',
        users: {
          'user-1': {
            displayName: 'new display name',
            fileName: 'new file name',
            searchName: 'new search name',
          },
        },
      })
      expect(res.status).to.equal(200)
      expect(res.body).toMatchSnapshot()
    })

    it('sends 204 if the show does not exist in store', async (ctx) => {
      // @ts-ignore
      getDocWhere.mockResolvedValue({ id: 'user-1' })
      // @ts-ignore
      getDoc.mockResolvedValue(undefined)

      const res = await ctx.request.put('/tv/shows/id')
      .set('api-key', 'user-1-api-key')
      .send({ show: {} })

      expect(res.status).to.equal(204)
    })
  })

  describe('DELETE /tv/shows/:id', () => {
    it('deletes the show if no other users reference it', async (ctx) => {
      const show = makeShow(1, [makeEpisode(1)], [1])

      // @ts-ignore
      getDocWhere.mockResolvedValue({ id: 'user-1' })
      // @ts-ignore
      getDoc.mockResolvedValue(show)

      const res = await ctx.request.delete('/tv/shows/id')
      .set('api-key', 'user-1-api-key')

      expect(deleteDoc).toBeCalledWith('shows/id')
      expect(res.status).to.equal(204)
    })

    it('removes the show from the user if another user references it', async (ctx) => {
      const show = makeShow(1, [makeEpisode(1)], [1, 2])

      // @ts-ignore
      getDocWhere.mockResolvedValue({ id: 'user-1' })
      // @ts-ignore
      getDoc.mockResolvedValue(show)

      const res = await ctx.request.delete('/tv/shows/id')
      .set('api-key', 'user-1-api-key')

      expect(updateDoc).toBeCalledWith('shows/id', {
        'user-2': {
          displayName: 'user-2 display name',
          fileName: 'user-2 file name',
          searchName: 'user-2 search name',
        },
      })
      expect(res.status).to.equal(204)
    })

    it('does nothing if show does not exist', async (ctx) => {
      // @ts-ignore
      getDocWhere.mockResolvedValue({ id: 'user-1' })
      // @ts-ignore
      getDoc.mockResolvedValue(undefined)

      const res = await ctx.request.delete('/tv/shows/id')
      .set('api-key', 'user-1-api-key')

      expect(deleteDoc).not.toBeCalled
      expect(updateDoc).not.toBeCalled
      expect(res.status).to.equal(204)
    })

    it('does nothing if show does not have current user', async (ctx) => {
      const show = makeShow(1, [makeEpisode(1)], [2])

      // @ts-ignore
      getDocWhere.mockResolvedValue({ id: 'user-1' })
      // @ts-ignore
      getDoc.mockResolvedValue(show)

      const res = await ctx.request.delete('/tv/shows/id')
      .set('api-key', 'user-1-api-key')

      expect(deleteDoc).not.toBeCalled
      expect(updateDoc).not.toBeCalled
      expect(res.status).to.equal(204)
    })
  })
})
