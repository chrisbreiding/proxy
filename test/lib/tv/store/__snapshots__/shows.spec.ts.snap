// Vitest Snapshot v1

exports[`lib/tv/store/shows > GET /tv/shows > returns shows from store for user 1`] = `
[
  {
    "displayName": "user-1 display name",
    "episodes": [
      {
        "title": "Episode 1",
      },
      {
        "title": "Episode 2",
      },
    ],
    "fileName": "user-1 file name",
    "id": "show-1-id",
    "searchName": "user-1 search name",
  },
  {
    "displayName": "user-1 display name",
    "episodes": [
      {
        "title": "Episode 3",
      },
      {
        "title": "Episode 4",
      },
    ],
    "fileName": "user-1 file name",
    "id": "show-2-id",
    "searchName": "user-1 search name",
  },
]
`;

exports[`lib/tv/store/shows > POST /tv/shows > adds the show to the user if it already exists 1`] = `
{
  "id": "show-3-id",
  "name": "Show 3 name",
  "users": {
    "user-1": {
      "displayName": "Show 3 name",
      "fileName": "Show 3 name",
      "searchName": "Show 3 name",
    },
    "user-2": {
      "displayName": "user-2 display name",
      "fileName": "user-2 file name",
      "searchName": "user-2 search name",
    },
  },
}
`;

exports[`lib/tv/store/shows > POST /tv/shows > gets the show data, saves it, and adds the show to the user if it does not already exist 1`] = `
{
  "id": "show-2-id",
  "name": "Show 2 name",
  "poster": "Show 2 poster",
  "status": "Ended",
  "users": {
    "user-1": {
      "displayName": "Show 2 name",
      "fileName": "Show 2 name",
      "searchName": "Show 2 name",
    },
  },
}
`;

exports[`lib/tv/store/shows > PUT /tv/shows/:id > updates the show and returns it 1`] = `
{
  "id": "show-1-id",
  "name": "Show 1 name",
  "users": {
    "user-1": {
      "displayName": "new display name",
      "fileName": "new file name",
      "searchName": "new search name",
    },
  },
}
`;
